File: finance-ui/src/app/services/auth.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface AuthResponse {
  token: string;
  email: string;
  name: string;
  userId: number;
}

@Injectable({
  providedIn: 'root',
})
export class AuthService {
  private readonly http = inject(HttpClient);

  private readonly baseUrl = '/auth';

  login(email: string, password: string): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(`${this.baseUrl}/login`, { email, password });
  }

  register(email: string, password: string, name: string, phone: string, role: string = 'USER'): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(`${this.baseUrl}/register`, { 
      email, 
      password, 
      fullName: name, 
      phoneNumber: phone, 
      role 
    });
  }

  saveSession(authData: AuthResponse): void {
    localStorage.setItem('auth_token', authData.token);
    localStorage.setItem('auth_email', authData.email);
    localStorage.setItem('auth_name', authData.name || '');
    localStorage.setItem('auth_user_id', authData.userId ? authData.userId.toString() : '');
  }

  getToken(): string | null {
    return localStorage.getItem('auth_token');
  }

  getEmail(): string | null {
    return localStorage.getItem('auth_email');
  }

  getName(): string | null {
    return localStorage.getItem('auth_name');
  }

  getUserId(): number | null {
    const id = localStorage.getItem('auth_user_id');
    return id ? parseInt(id, 10) : null;
  }

  clearSession(): void {
    localStorage.removeItem('auth_token');
    localStorage.removeItem('auth_email');
    localStorage.removeItem('auth_name');
    localStorage.removeItem('auth_user_id');
  }

  isLoggedIn(): boolean {
    return !!this.getToken();
  }
}

File: my_finance_mobile_ui/lib/services/auth_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../models/models.dart';
import 'api_config.dart';

class AuthService {
  final String _authUrl = '${ApiConfig.baseUrl}/auth';

  // Perform login
  Future<AuthResponse> login(String email, String password) async {
    final response = await http.post(
      Uri.parse('$_authUrl/login'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'email': email,
        'password': password,
      }),
    );

    if (response.statusCode == 200) {
      final decodedJson = jsonDecode(utf8.decode(response.bodyBytes));
      final authResponse = AuthResponse.fromJson(decodedJson);
      await saveSession(authResponse);
      return authResponse;
    } else {
      String errorMessage = 'Giriş yapılamadı!';
      try {
        final errorBody = jsonDecode(utf8.decode(response.bodyBytes));
        errorMessage = errorBody['message'] ?? errorMessage;
      } catch (_) {}
      throw Exception(errorMessage);
    }
  }

  // Perform registration
  Future<AuthResponse> register({
    required String email,
    required String password,
    required String fullName,
    required String phoneNumber,
    String role = 'USER',
  }) async {
    final response = await http.post(
      Uri.parse('$_authUrl/register'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'email': email,
        'password': password,
        'fullName': fullName,
        'phoneNumber': phoneNumber,
        'role': role,
      }),
    );

    if (response.statusCode == 200) {
      final decodedJson = jsonDecode(utf8.decode(response.bodyBytes));
      final authResponse = AuthResponse.fromJson(decodedJson);
      await saveSession(authResponse);
      return authResponse;
    } else {
      String errorMessage = 'Kayıt olunamadı!';
      try {
        final errorBody = jsonDecode(utf8.decode(response.bodyBytes));
        errorMessage = errorBody['message'] ?? errorMessage;
      } catch (_) {}
      throw Exception(errorMessage);
    }
  }

  // Save session info
  Future<void> saveSession(AuthResponse authResponse) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('auth_token', authResponse.token);
    await prefs.setString('auth_email', authResponse.email);
    await prefs.setString('auth_name', authResponse.name);
    await prefs.setInt('auth_user_id', authResponse.userId);
  }

  // Retrieve token
  Future<String?> getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('auth_token');
  }

  // Retrieve email
  Future<String?> getEmail() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('auth_email');
  }

  // Retrieve name
  Future<String?> getName() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('auth_name');
  }

  // Retrieve user ID
  Future<int?> getUserId() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getInt('auth_user_id');
  }

  // Clear session (logout)
  Future<void> clearSession() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('auth_token');
    await prefs.remove('auth_email');
    await prefs.remove('auth_name');
    await prefs.remove('auth_user_id');
  }

  // Check login status
  Future<bool> isLoggedIn() async {
    final token = await getToken();
    return token != null && token.isNotEmpty;
  }
}
